### ISSUE TITLE
Server purchase research

### ISSUE THREAD

**Issue by SimonBiggs** [dicomrtai/dicomrtai#2]
https://forums.fast.ai/t/recommendations-on-new-2-x-rtx-3090-setup/78202/268?u=simonbiggs

**Comment by SimonBiggs** [comment]
Also, heads up @matthewdeancooper @pchlap, I'm looking at building a machine learning server with its computational resources exposed to the general public for use as an auto-segmentation web-API.

**Comment by matthewdeancooper** [comment]
2 x 3090s ? Going to need a bigger PSU 😂.

**Comment by SimonBiggs** [comment]
Hehe, 1200W should be sufficient. I'll also be slightly under-clocking the machine. Others have shown that significantly less wall power draw can be achieved for only a 5% reduction in performance.

**Comment by matthewdeancooper** [comment]
I wonder if a business like https://www.rev.ai/ could ever work in the RT segmentation space? Essentially, people pay rev.ai for speech to text conversion and they are the best in the business (as far as I am aware). The processing is done first by a model and then corrected by a human. Cool part is that all data that is labelled for customers can go straight back into improving their model. In a way, they are being paid to make their own training data.

**Comment by SimonBiggs** [comment]
> The processing is done first by a model and then corrected by a human. Cool part is that all data that is labelled for customers can go straight back into improving their model.

That's essentially the idea here. Main difference being, by offering QA services the users are providing the training data. As in, users submit data to be QAed, and then the model provides the QA result. But the original data that was provided by the user can be fed back into the model to make the QA process better for everyone.

**Comment by matthewdeancooper** [comment]
"Back into the os community" would be a good way to fund some development on the pymedphys front. I like the idea, what do you need help with?

**Comment by SimonBiggs** [comment]
> I like the idea, what do you need help with?

I like the enthusiasm :slightly_smiling_face:.

---

I've been mulling over a network infrastructure that has variable inputs and outputs. The goal being, have the input be able to be MRI (multiple types), PET, and/or CT (either one, some, or all), and have the output be able to be variable also, but have a well trained, generic, UNet back-bone.

Have the weights on the first layer of the network be modality specific, have the weights on the last layer be contour specific.

The idea would be to train the network with a standard set of inputs and outputs initially. Then, swap out just the outputs masks, along with their weights. Initially freeze the weights of the entire U-Net, except the weights connecting to the output. Train for a while (it won't work very well initially). But then, iteratively swap between each output type between epochs. Then do the same with the inputs, and their corresponding weights.

---

So, goal is, have every bit of data feed into the generic backbone, instead of having separate models for every contour type and imaging modality.

---

For this to work, the data input needs to have a consistent format that will work for all contour and imaging types. I've been mulling on that too, and I really like the concept of patch based training. So, the idea is, have the data that is being input into the model be a 128x128x128 voxel cube. This cube will have random rotation along all three axes, and random position. The voxel size will be 1x1x1 mm. These voxels will be created by running linear interpolation within the imaging space. During training the cube position and rotation is selected at random, but then some of the cubes will be discarded if they don't have a specific contour within them. The contour(s) being "focused on" can be guided by a range of metrics with the goal to direct the training and minimise the effects of class imbalance during training.

---

My justification for going down this route, is I want the model design to be optimised for being able to pick up new contour types. That way, a user can provide a minimal "fine-tuning" dataset, and then from their the model can begin to undergo QA on the user's contours. Also, I want to build a degree of robustness into its backbone through generalisation, so that each contour type benefits from the training and data available for the other contour types.

---

So after that precursor... my first task that I thought would be best would be to take a large public dataset such as the Kidney dataset (https://kits19.grand-challenge.org/home/), and begin making a pipeline for that dataset that creates these cubes. Potentially I could start with 64x64x64 interpolated cubes (or even 32x32x32) cubes. And then see if this custom training idea works. I can do that by flipping between the kidney and another contour type (say the pancreas dataset https://wiki.cancerimagingarchive.net/display/Public/Pancreas-CT#225140400c26eab54502412cbbd0e1c0fddd917b) while having the last layer of the network swap out with the contour type, and then tweak the training process.

With more minds in the game though, a better first task is brainstorming. 

---

Also, parallel tasks, are brainstorming on the server API, user data submission and client side anonymisation. Setting up a cloud server, and secure data hosting. Setting up a means for users to submit large data via the API. Make that seamless. Buying and building the local machine learning server. Have it expose a private API just for use by the cloud server which is designed for triggering the local machine learning machine to undergo training/fine tuning/inference on a specified dataset. Potentially offloading inference to a GPU in the cloud as that may be a more cost effective use of resources.

... So there are "machine learning investigation tasks", but also "server infrastructure tasks", and lastly "graphical user interface, and user experience" tasks. All of which need further brainstorming and thought.

So... the real first task is to undergo brainstorming. Here's as good a place as any to do that.

**Comment by pchlap** [comment]
This all sounds super interesting @SimonBiggs 

A lot of overlap here with what I am working on my PhD and some of the tools which I have been working on integrating into our clinical systems (if you remember my EPSM pres). I think it would definitely make sense to try to combine our efforts on this.

Keen for a call sometime in the coming week?

**Comment by matthewdeancooper** [comment]
I've put aside a few hours tonight to had a google hunt.

So in summary:

A 3D UNet architecture trained on x**3 patches for segmentation:
1) Backbone that has learned segmentation kernels in a general sense and is modality/contour type invariant
2) Multiple output layers that are contour specific and invariant to the modality type
3) Multiple initial layers that are modality specific and invariant to the contour type

Back end training:
1) Learn a general segmentation task to create an initial model
2) Freeze everything except the final layer. Save independently trained output-weights for each output type
3) Do the same thing for the input layer but for modality type, exposing each to all output/output-weight pairs

User training:
1) Select the correct input/output weights for their task.
2) Train the entire network for their task as the optimal transformation is unlikely invariant to the above.

Am i on the right track here before i dig any deeper?

**Comment by SimonBiggs** [comment]
> Am i on the right track here before i dig any deeper?

Yup, exactly the right track. Only clarification is on that last point. When the entire model is retrained alternate between the new user data and random samples from the previous data making sure that the fine tuning being undergone for the new user data doesn't degrade the model for the previous data. The new to old ratio would be tweaked so that results on old data doesn't degrade but the results on the new data continues to improve.

Also, I'd suspect the training for new user data would be in two stages. First would be to train with just the input and output layers being trainable first (to see if the generic backbone is close), and then, after that's no longer improving, then allow the entire model to be trained (while alternating with old data).

**Comment by SimonBiggs** [comment]
> A lot of overlap here with what I am working on my PhD and some of the tools which I have been working on integrating into our clinical systems
> ...
> Keen for a call sometime in the coming week?

Yup, that would be awesome. I'm still in recovery mode, so let's give it a few more days until I'm back on top 🙂

**Comment by pchlap** [comment]
> > A lot of overlap here with what I am working on my PhD and some of the tools which I have been working on integrating into our clinical systems
> > ...
> > Keen for a call sometime in the coming week?
> 
> Yup, that would be awesome. I'm still in recovery mode, so let's give it a few more days until I'm back on top 🙂

Absolutely. Get well soon!

**Comment by SimonBiggs** [comment]
As a heads up @pchlap, when mulling over how to interface client software with the public server I have been having the following thoughts:

When it comes to the API side of things I'm leaning towards utilising SFTP for transferring the large DICOM files to and from the user, and then using protocol buffers as the key interface technology between client side software and the public facing server.

Thoughts are, the client side software that utilises the API would generate its own specific SSH public/private keys where the public key gets associated with their API key on the https://dicomrt.ai cloud account management server. Then using those SSH credentials they submit the large files to the server via SFTP. And everything is set up such that each API key can only access the files that have been submitted by the SSH credentials that have been associated with that API key.

So the web-API would come in a two parts. The SFTP file transfer API for both sending data, and receiving results, and secondly the protocol buffer API for interfacing with the server and making the required requests/commands.

**Comment by SimonBiggs** [comment]
Also, the client side software would anonymised the DICOM files before submission.

**Comment by SimonBiggs** [comment]
Also... Given how complicated money makes things... And how it can introduce expectations around what should be given around open source work... I would like to make it clear that all software around this project is going to be open source, other's are free to use it at no cost. But in return software being shared and contributed to the dicomrtai project should not be undergone under an unspoken expectation of being remunerated for it in the future. If it is wildly successful in the future I would certainly expect that machine learning and software engineers would be hired, but that would be being paid for work done as part of that hiring agreement.

On the topic of forming some kind of partnership, more than happy to build the code with others in such a way that it is maximally beneficial to other projects also, but on the partnership/company/commercial side, if this does ever go down that route, we've already had a team in mind that optimises for new and diversified skills that don't already exist within the group. As in... The group is chosen and designed to be as small as possible and in such a way where each added person brings something orthogonal to everyone else already in the group. So, in saying that, so that everyone's expectations are clear with this sort of thing when money is involved, any partnership/company that might form would likely involve @Matthew-Jennings, @darcymason, and one other person who I won't name yet as they are not entirely decided on whether or not they are in a position to join.

Normally the agreement around open source is "everyone gives code, and expects only that others be as willing to share", I wanted to lay out these expectations nice and clear as I am concerned that there maybe unspoken expectations creeping into the mix, and I don't want people to feel hard done by, with their expectations not met, down the road.

**Comment by matthewdeancooper** [comment]
This seems like a common strategy for people who are mining crypto to reduce their power bills.

**Comment by pchlap** [comment]
Thats sounds very exciting @SimonBiggs... Not sure if you've seen this or if relevant for the sort of thing you are describing here, but here is a funding opportunity specifically for open source projects: https://chanzuckerberg.com/rfa/essential-open-source-software-for-science/

**Comment by matthewdeancooper** [comment]
> 
> 
> Normally the agreement around open source is "everyone gives code, and expects only that others be as willing to share", I wanted to lay out these expectations nice and clear as I am concerned that there may be unspoken expectations creeping into the mix, and I don't want people to feel hard done by, with their expectations not met, down the road.

The rev.ai mention was more-so a suggestion/template on how to expand the service beyond an open source api/model. For instance, by offering human correction on-top of the model  predictions - which would necessitate a cost and a contract with the corrector. 

The main reason for this suggestion is that it is a way to implement active learning where any predictions can be corrected to become accurate labels for training. In the case of rev.ai, I found this to be an interesting way to improve the model and the utility of the service. Healthcare seems to be missing this active learning step as all predictions are likely corrected prior to clinical use, and so there is a missed opportunity to feed this corrective manpower back into improving predictions.

Good to get the project goals out in the open though!

**Comment by Matthew-Jennings** [comment]
A6000 vs 2x 3090s?

**Comment by SimonBiggs** [comment]
> A6000 vs 2x 3090s?

Number of CUDA cores and cost. 1x 3090 is roughly comparable to 1x A6000 in training speed. Nevertheless, having a single A6000 will have enough GPU RAM to still be able to train on the same models as 2x 3090, albeit ~half the speed.
